File: objc/Tests/TestM97LongReadProfile.m
// TestM97LongReadProfile.m — M97 long-read profile: the @read_role
// attribute, the QUALITY_BINNED platform guard, and the REF_DIFF_V2
// slice_bytes byte budget through the writer paths.
//
// Mirrors:
//   python/tests/test_m97_long_read_profile.py
//   java/src/test/java/.../M97LongReadProfileTest.java
//
// SPDX-License-Identifier: LGPL-3.0-or-later

#import <Foundation/Foundation.h>
#import <hdf5.h>
#import "Testing.h"
#import "Dataset/TTIOSpectralDataset.h"
#import "Dataset/TTIOSpectralDataset+GenomicWrite.h"
#import "Genomics/TTIOAlignedRead.h"
#import "Genomics/TTIOWrittenGenomicRun.h"
#import "Genomics/TTIOGenomicRun.h"
#import "Genomics/TTIOGenomicStreamWriter.h"
#import "Genomics/TTIOGenomicBlocks.h"
#import "Genomics/TTIOGenomicWriteContext.h"
#import "Codecs/TTIOQuality.h"
#import "Providers/TTIOMemoryProvider.h"
#import "Providers/TTIOProviderRegistry.h"
#import "Providers/TTIOStorageProtocols.h"
#import "ValueClasses/TTIOEnums.h"
#include <unistd.h>

static NSString *m97TmpPath(NSString *tag)
{
    return [NSString stringWithFormat:@"/tmp/ttio_m97_%d_%@.tio",
            (int)getpid(), tag];
}

static void m97Rm(NSString *p)
{
    [[NSFileManager defaultManager] removeItemAtPath:p error:NULL];
}

/** Reference: 1000 x "ACGTACGTAC" (10 000 bp). */
static NSData *m97Ref(void)
{
    NSData *unit = [@"ACGTACGTAC" dataUsingEncoding:NSASCIIStringEncoding];
    NSMutableData *d = [NSMutableData data];
    for (int i = 0; i < 1000; i++) [d appendData:unit];
    return d;
}

/** An aligned 40-read single-chromosome run, 10 bp per read, positions
 *  i*20 + 1, cigar "10M". Legacy whole-channel layout so the refdiff_v2
 *  blob sits at a fixed dataset path. */
static TTIOWrittenGenomicRun *m97MakeAlignedRun(NSString *platform)
{
    NSUInteger n = 40;
    NSUInteger L = 10;

    NSMutableData *seqData  = [NSMutableData dataWithCapacity:n * L];
    NSMutableData *qualData = [NSMutableData dataWithLength:n * L];
    for (NSUInteger i = 0; i < n; i++)
        [seqData appendBytes:"ACGTACGTAC" length:L];
    memset(qualData.mutableBytes, 30, n * L);

    NSMutableData *posD = [NSMutableData dataWithLength:n * sizeof(int64_t)];
    NSMutableData *mqD  = [NSMutableData dataWithLength:n * sizeof(uint8_t)];
    NSMutableData *flD  = [NSMutableData dataWithLength:n * sizeof(uint32_t)];
    NSMutableData *offD = [NSMutableData dataWithLength:n * sizeof(uint64_t)];
    NSMutableData *lenD = [NSMutableData dataWithLength:n * sizeof(uint32_t)];
    NSMutableData *mpD  = [NSMutableData dataWithLength:n * sizeof(int64_t)];
    NSMutableData *tlD  = [NSMutableData dataWithLength:n * sizeof(int32_t)];

    int64_t  *pos = (int64_t  *)posD.mutableBytes;
    uint8_t  *mq  = (uint8_t  *)mqD.mutableBytes;
    uint32_t *fl  = (uint32_t *)flD.mutableBytes;
    uint64_t *off = (uint64_t *)offD.mutableBytes;
    uint32_t *len = (uint32_t *)lenD.mutableBytes;
    int64_t  *mp  = (int64_t  *)mpD.mutableBytes;
    int32_t  *tl  = (int32_t  *)tlD.mutableBytes;

    for (NSUInteger i = 0; i < n; i++) {
        pos[i] = (int64_t)(i * 20 + 1);
        mq[i]  = 60;
        fl[i]  = 0;
        off[i] = i * L;
        len[i] = (uint32_t)L;
        mp[i]  = -1;
        tl[i]  = 0;
    }

    NSMutableArray *chroms     = [NSMutableArray array];
    NSMutableArray *mateChroms = [NSMutableArray array];
    NSMutableArray *cigars     = [NSMutableArray array];
    NSMutableArray *readNames  = [NSMutableArray array];
    for (NSUInteger i = 0; i < n; i++) {
        [chroms     addObject:@"22"];
        [mateChroms addObject:@"*"];
        [cigars     addObject:@"10M"];
        [readNames  addObject:[NSString stringWithFormat:@"r%lu", (unsigned long)i]];
    }

    TTIOWrittenGenomicRun *run =
        [([[TTIOWrittenGenomicRun alloc]
         initWithAcquisitionMode:TTIOAcquisitionModeGenomicWGS
                    referenceUri:@"GRCh38"
                        platform:platform
                      sampleName:@"HG002"
                       positions:posD
                mappingQualities:mqD
                           flags:flD
                       sequences:seqData
                       qualities:qualData
                         offsets:offD
                         lengths:lenD
                          cigars:cigars
                       readNames:readNames
                 mateChromosomes:mateChroms
                   matePositions:mpD
                 templateLengths:tlD
                     chromosomes:chroms
               signalCompression:TTIOCompressionZlib
            signalCodecOverrides:@{}]) copyWithOptLegacyWholeChannel:YES];

    run.referenceChromSeqs = @{ @"22": m97Ref() };
    run.embedReference = YES;
    return run;
}

/** Raw refdiff_v2 blob bytes of the legacy whole-channel file at
 *  @p cpath, or nil. */
static NSData *m97LegacyRefDiffBlob(const char *cpath)
{
    hid_t f = H5Fopen(cpath, H5F_ACC_RDONLY, H5P_DEFAULT);
    if (f < 0) return nil;
    hid_t did = H5Dopen2(
        f, "study/genomic_runs/genomic_0001/signal_channels/sequences/refdiff_v2",
        H5P_DEFAULT);
    if (did < 0) { H5Fclose(f); return nil; }
    hid_t sp = H5Dget_space(did);
    hssize_t n = H5Sget_simple_extent_npoints(sp);
    H5Sclose(sp);
    NSMutableData *out = nil;
    if (n > 0) {
        out = [NSMutableData dataWithLength:(NSUInteger)n];
        if (H5Dread(did, H5T_NATIVE_UINT8, H5S_ALL, H5S_ALL, H5P_DEFAULT,
                    out.mutableBytes) < 0)
            out = nil;
    }
    H5Dclose(did);
    H5Fclose(f);
    return out;
}

static uint32_t m97NSlices(NSData *blob)
{
    if (blob.length < 12) return 0;
    const uint8_t *b = (const uint8_t *)blob.bytes;
    return (uint32_t)b[8] | ((uint32_t)b[9] << 8)
         | ((uint32_t)b[10] << 16) | ((uint32_t)b[11] << 24);
}

// ── QUALITY_BINNED platform guard ─────────────────────────────────

static void m97PlatformPredicate(void)
{
    PASS(TTIOQualityBinnedAllowedForPlatform(nil),
         "M97 guard: nil platform allowed");
    PASS(TTIOQualityBinnedAllowedForPlatform(@""),
         "M97 guard: empty platform allowed");
    PASS(TTIOQualityBinnedAllowedForPlatform(@"ILLUMINA"),
         "M97 guard: ILLUMINA allowed");
    PASS(TTIOQualityBinnedAllowedForPlatform(@"IONTORRENT"),
         "M97 guard: IONTORRENT allowed (ont only as a token)");
    PASS(!TTIOQualityBinnedAllowedForPlatform(@"ONT"),
         "M97 guard: ONT rejected");
    PASS(!TTIOQualityBinnedAllowedForPlatform(@"PacBio HiFi"),
         "M97 guard: PacBio HiFi rejected");
    PASS(!TTIOQualityBinnedAllowedForPlatform(@"HIFI"),
         "M97 guard: HIFI rejected");
    PASS(!TTIOQualityBinnedAllowedForPlatform(@"Oxford Nanopore"),
         "M97 guard: Oxford Nanopore rejected");
}

static void m97QualityBinnedGuard(void)
{
    // writeMinimal path: an ONT run with a QUALITY_BINNED qualities
    // override raises before anything is written.
    TTIOWrittenGenomicRun *ont =
        [m97MakeAlignedRun(@"ONT") copyWithSignalCodecOverrides:
            @{@"qualities": @(TTIOCompressionQualityBinned)}];
    NSString *path = m97TmpPath(@"guard");
    m97Rm(path);
    BOOL raised = NO;
    @try {
        NSError *err = nil;
        [TTIOSpectralDataset writeMinimalToPath:path
                                          title:@"M97Guard"
                            isaInvestigationId:@"ISA-M97"
                                        msRuns:@{}
                                    genomicRuns:@{@"genomic_0001": ont}
                                identifications:nil
                                quantifications:nil
                              provenanceRecords:nil
                                          error:&err];
    } @catch (NSException *e) {
        raised = [e.name isEqualToString:NSInvalidArgumentException];
    }
    PASS(raised, "M97 guard: writeMinimal rejects QUALITY_BINNED on ONT");
    m97Rm(path);

    // Stream-writer path: rejected at init.
    NSString *url = [NSString stringWithFormat:@"memory://m97-guard-%d", (int)getpid()];
    [TTIOMemoryProvider discardStore:url];
    id<TTIOStorageProvider> mem = [[TTIOProviderRegistry sharedRegistry]
        openURL:url mode:TTIOStorageOpenModeCreate provider:@"memory" error:NULL];
    id<TTIOStorageGroup> study =
        [[mem rootGroupWithError:NULL] createGroupNamed:@"study" error:NULL];
    TTIOGenomicStreamWriterOptions *o = [TTIOGenomicStreamWriterOptions defaultOptions];
    o.platform = @"PacBio HiFi";
    o.signalCodecOverrides = @{@"qualities": @(TTIOCompressionQualityBinned)};
    raised = NO;
    @try {
        (void)[[TTIOGenomicStreamWriter alloc] initWithStudyGroup:study
                                                          runName:@"g"
                                                          options:o];
    } @catch (NSException *e) {
        raised = [e.name isEqualToString:NSInvalidArgumentException];
    }
    PASS(raised, "M97 guard: stream writer rejects QUALITY_BINNED on HiFi");

    // The same override on a short-read platform is accepted.
    o = [TTIOGenomicStreamWriterOptions defaultOptions];
    o.platform = @"ILLUMINA";
    o.signalCodecOverrides = @{@"qualities": @(TTIOCompressionQualityBinned)};
    BOOL ok = YES;
    @try {
        (void)[[TTIOGenomicStreamWriter alloc] initWithStudyGroup:study
                                                          runName:@"g2"
                                                          options:o];
    } @catch (NSException *e) {
        ok = NO;
    }
    PASS(ok, "M97 guard: QUALITY_BINNED accepted on ILLUMINA");
}

// ── @read_role round-trip ─────────────────────────────────────────

static void m97ReadRoleRoundTrip(void)
{
    NSString *path = m97TmpPath(@"role");
    m97Rm(path);
    TTIOWrittenGenomicRun *run = m97MakeAlignedRun(@"PacBio HiFi");
    run.readRole = @"hifi";
    NSError *err = nil;
    PASS([TTIOSpectralDataset writeMinimalToPath:path
                                           title:@"M97Role"
                             isaInvestigationId:@"ISA-M97"
                                         msRuns:@{}
                                     genomicRuns:@{@"genomic_0001": run}
                                 identifications:nil
                                 quantifications:nil
                               provenanceRecords:nil
                                           error:&err],
         "M97 read_role: write succeeds (%s)",
         [[err localizedDescription] UTF8String] ?: "");

    TTIOSpectralDataset *ds = [TTIOSpectralDataset readFromFilePath:path error:&err];
    TTIOGenomicRun *gr = ds.genomicRuns[@"genomic_0001"];
    PASS([gr.readRole isEqualToString:@"hifi"],
         "M97 read_role: accessor returns 'hifi' (got %@)",
         gr.readRole ?: @"<nil>");
    [ds closeFile];
    m97Rm(path);

    // Absent attribute (pre-M97 file shape) reads back as nil.
    m97Rm(path);
    TTIOWrittenGenomicRun *plain = m97MakeAlignedRun(@"ILLUMINA");
    PASS([TTIOSpectralDataset writeMinimalToPath:path
                                           title:@"M97RoleAbsent"
                             isaInvestigationId:@"ISA-M97"
                                         msRuns:@{}
                                     genomicRuns:@{@"genomic_0001": plain}
                                 identifications:nil
                                 quantifications:nil
                               provenanceRecords:nil
                                           error:&err],
         "M97 read_role: role-less write succeeds");
    ds = [TTIOSpectralDataset readFromFilePath:path error:&err];
    gr = ds.genomicRuns[@"genomic_0001"];
    PASS(gr != nil && gr.readRole == nil,
         "M97 read_role: absent attribute reads back nil");
    [ds closeFile];
    m97Rm(path);

    // blocks_v1 path: the stream writer stamps the attribute from its
    // options.
    NSString *url = [NSString stringWithFormat:@"memory://m97-role-%d", (int)getpid()];
    [TTIOMemoryProvider discardStore:url];
    id<TTIOStorageProvider> mem = [[TTIOProviderRegistry sharedRegistry]
        openURL:url mode:TTIOStorageOpenModeCreate provider:@"memory" error:NULL];
    id<TTIOStorageGroup> study =
        [[mem rootGroupWithError:NULL] createGroupNamed:@"study" error:NULL];
    TTIOWrittenGenomicRun *blockRun = m97MakeAlignedRun(@"ONT");
    blockRun = [blockRun copyWithOptLegacyWholeChannel:NO];
    blockRun.readRole = @"ont_ul";
    TTIOGenomicStreamWriterOptions *o =
        [TTIOGenomicStreamWriterOptions optionsFromRun:blockRun];
    TTIOGenomicStreamWriter *w =
        [[TTIOGenomicStreamWriter alloc] initWithStudyGroup:study
                                                    runName:@"g"
                                                    options:o];
    PASS([w appendBatch:blockRun error:&err], "M97 read_role: blocks append");
    PASS([w close:&err], "M97 read_role: blocks close");
    id<TTIOStorageGroup> rg =
        [[study openGroupNamed:@"genomic_runs" error:NULL] openGroupNamed:@"g"
                                                                    error:NULL];
    NSString *role = [rg attributeValueForName:@"read_role" error:NULL];
    PASS([role isEqualToString:@"ont_ul"],
         "M97 read_role: blocks_v1 run attribute is 'ont_ul' (got %@)",
         role ?: @"<nil>");
}

// ── REF_DIFF_V2 slice_bytes through the writer ────────────────────

static void m97SliceBytesThroughWriter(void)
{
    // The blocks encoder carries refDiffSliceBytes into the refdiff
    // codec: a 100-base budget over 40 x 10 bp reads makes 4 slices
    // where the default makes 1.
    NSError *err = nil;
    TTIOWrittenGenomicRun *plain =
        [m97MakeAlignedRun(@"PacBio HiFi") copyWithOptLegacyWholeChannel:NO];
    TTIOBlockBlobs *def = [TTIOGenomicBlocks encodeBlock:plain
                                                 context:[TTIOGenomicWriteContext none]
                                                   error:&err];
    PASS(def != nil, "M97 slice_bytes: default block encode (%s)",
         [[err localizedDescription] UTF8String] ?: "");
    if (!def) return;
    PASS([def.codecs[@"sequences"] integerValue] == TTIOCompressionRefDiffV2,
         "M97 slice_bytes: sequences channel is REF_DIFF_V2");
    PASS(m97NSlices(def.blobs[@"sequences"]) == 1,
         "M97 slice_bytes: default is 1 slice (got %u)",
         m97NSlices(def.blobs[@"sequences"]));

    TTIOWrittenGenomicRun *budgeted =
        [m97MakeAlignedRun(@"PacBio HiFi") copyWithOptLegacyWholeChannel:NO];
    budgeted.refDiffSliceBytes = 100;
    TTIOBlockBlobs *bb = [TTIOGenomicBlocks encodeBlock:budgeted
                                                context:[TTIOGenomicWriteContext none]
                                                  error:&err];
    PASS(bb != nil, "M97 slice_bytes: budgeted block encode");
    if (!bb) return;
    uint32_t nSlices = m97NSlices(bb.blobs[@"sequences"]);
    PASS(nSlices == 4,
         "M97 slice_bytes: 100-base budget makes 4 slices (got %u)", nSlices);

    // Legacy whole-channel writeMinimal path: same knob, and the file
    // round-trips through the ordinary reader.
    NSString *path = m97TmpPath(@"slice");
    m97Rm(path);
    TTIOWrittenGenomicRun *legacy = m97MakeAlignedRun(@"PacBio HiFi");
    legacy.refDiffSliceBytes = 100;
    PASS([TTIOSpectralDataset writeMinimalToPath:path
                                           title:@"M97Slice"
                             isaInvestigationId:@"ISA-M97"
                                         msRuns:@{}
                                     genomicRuns:@{@"genomic_0001": legacy}
                                 identifications:nil
                                 quantifications:nil
                               provenanceRecords:nil
                                           error:&err],
         "M97 slice_bytes: legacy write succeeds (%s)",
         [[err localizedDescription] UTF8String] ?: "");
    NSData *legacyBlob = m97LegacyRefDiffBlob([path fileSystemRepresentation]);
    PASS(m97NSlices(legacyBlob) == 4,
         "M97 slice_bytes: legacy HDF5 blob carries 4 slices (got %u)",
         m97NSlices(legacyBlob));
    TTIOSpectralDataset *ds = [TTIOSpectralDataset readFromFilePath:path error:&err];
    TTIOGenomicRun *gr = ds.genomicRuns[@"genomic_0001"];
    BOOL allCorrect = (gr != nil && gr.readCount == 40);
    for (NSUInteger i = 0; allCorrect && i < 40; i++) {
        TTIOAlignedRead *r = [gr readAtIndex:i error:&err];
        if (!r || ![r.sequence isEqualToString:@"ACGTACGTAC"]) allCorrect = NO;
    }
    PASS(allCorrect,
         "M97 slice_bytes: budgeted legacy file round-trips 40 reads");
    [ds closeFile];
    m97Rm(path);
}

void testM97LongReadProfile(void);
void testM97LongReadProfile(void)
{
    m97PlatformPredicate();
    m97QualityBinnedGuard();
    m97ReadRoleRoundTrip();
    m97SliceBytesThroughWriter();
}

File: python/tests/test_m97_long_read_profile.py
"""M97 long-read profile tests.

The ``@read_role`` attribute, the QUALITY_BINNED platform guard, and
the REF_DIFF_V2 ``slice_bytes`` byte budget — at the codec level and
through both writer paths (legacy whole-channel and blocks_v1).

Mirrors:
    objc/Tests/TestM97LongReadProfile.m
    java/src/test/java/.../M97LongReadProfileTest.java
"""
from __future__ import annotations

import struct
from pathlib import Path

import numpy as np
import pytest

from ttio.codecs import quality
from ttio.codecs import ref_diff_v2 as rdv2
from ttio.enums import Compression

if not rdv2.HAVE_NATIVE_LIB:
    pytest.skip(
        "requires native libttio_rans.so via TTIO_RANS_LIB_PATH",
        allow_module_level=True,
    )


# ---------------------------------------------------------------------------
# fixtures
# ---------------------------------------------------------------------------

N = 40
READ_LEN = 10


def _ref_bytes(length: int = 10_000) -> bytes:
    return (b"ACGTACGTAC" * (length // 10 + 1))[:length]


def _build_run(platform: str = "ILLUMINA", **extra):
    """An aligned 40-read single-chromosome run, 10 bp per read."""
    from ttio.written_genomic_run import WrittenGenomicRun

    positions = (np.arange(N) * 20 + 1).astype(np.int64)
    ref = _ref_bytes()
    seqs = bytearray()
    for i in range(N):
        start = int(positions[i]) - 1
        seqs.extend(ref[start:start + READ_LEN])

    return WrittenGenomicRun(
        acquisition_mode=7,
        reference_uri="GRCh38.m97_test",
        platform=platform,
        sample_name="HG002",
        positions=positions,
        mapping_qualities=np.full(N, 60, dtype=np.uint8),
        flags=np.zeros(N, dtype=np.uint32),
        sequences=np.frombuffer(bytes(seqs), dtype=np.uint8),
        qualities=np.frombuffer(bytes([30] * (N * READ_LEN)), dtype=np.uint8),
        offsets=np.arange(N, dtype=np.uint64) * READ_LEN,
        lengths=np.full(N, READ_LEN, dtype=np.uint32),
        cigars=[f"{READ_LEN}M"] * N,
        read_names=[f"r{i}" for i in range(N)],
        mate_chromosomes=["*"] * N,
        mate_positions=np.zeros(N, dtype=np.int64),
        template_lengths=np.zeros(N, dtype=np.int32),
        chromosomes=["22"] * N,
        reference_chrom_seqs={"22": ref},
        embed_reference=True,
        **extra,
    )


def _write_run(tmp_path: Path, run, fname: str = "m97.tio") -> Path:
    from ttio.spectral_dataset import SpectralDataset

    out = tmp_path / fname
    SpectralDataset.write_minimal(
        out,
        title="m97_test",
        isa_investigation_id="M97",
        runs={},
        genomic_runs={"r0": run},
    )
    return out


def _n_slices(blob: bytes) -> int:
    """The outer header's n_slices (u32 LE at offset 8)."""
    return struct.unpack_from("<I", blob, 8)[0]


# ---------------------------------------------------------------------------
# QUALITY_BINNED platform guard
# ---------------------------------------------------------------------------

def test_binned_allowed_for_platform():
    assert quality.binned_allowed_for_platform(None)
    assert quality.binned_allowed_for_platform("")
    assert quality.binned_allowed_for_platform("ILLUMINA")
    # ont only counts as a whole token — IONTORRENT contains it.
    assert quality.binned_allowed_for_platform("IONTORRENT")
    assert not quality.binned_allowed_for_platform("ONT")
    assert not quality.binned_allowed_for_platform("PacBio HiFi")
    assert not quality.binned_allowed_for_platform("HIFI")
    assert not quality.binned_allowed_for_platform("Oxford Nanopore")


def test_quality_binned_guard_write_minimal(tmp_path: Path):
    run = _build_run(
        platform="ONT",
        opt_legacy_whole_channel=True,
        signal_codec_overrides={"qualities": Compression.QUALITY_BINNED},
    )
    with pytest.raises(ValueError, match="QUALITY_BINNED"):
        _write_run(tmp_path, run)

    # The same override on a short-read platform writes fine.
    ok = _build_run(
        platform="ILLUMINA",
        opt_legacy_whole_channel=True,
        signal_codec_overrides={"qualities": Compression.QUALITY_BINNED},
    )
    assert _write_run(tmp_path, ok, "ok.tio").exists()


def test_quality_binned_guard_stream_writer(tmp_path: Path):
    from ttio.genomic.stream_writer import GenomicStreamWriter
    from ttio.spectral_dataset import SpectralDataset

    out = tmp_path / "guard.tio"
    SpectralDataset.write_minimal(out, title="", isa_investigation_id="",
                                  runs={})
    with SpectralDataset.open(out, writable=True) as ds:
        with pytest.raises(ValueError, match="QUALITY_BINNED"):
            GenomicStreamWriter(
                ds.study_group, "g",
                acquisition_mode=7, reference_uri="", platform="PacBio HiFi",
                sample_name="",
                signal_codec_overrides={
                    "qualities": Compression.QUALITY_BINNED},
            )


# ---------------------------------------------------------------------------
# @read_role round-trip
# ---------------------------------------------------------------------------

def test_read_role_round_trip_legacy(tmp_path: Path):
    from ttio.spectral_dataset import SpectralDataset

    run = _build_run(platform="PacBio HiFi",
                     opt_legacy_whole_channel=True, read_role="hifi")
    out = _write_run(tmp_path, run)
    with SpectralDataset.open(out) as ds:
        assert ds.genomic_runs["r0"].read_role == "hifi"

    plain = _build_run(opt_legacy_whole_channel=True)
    out2 = _write_run(tmp_path, plain, "plain.tio")
    with SpectralDataset.open(out2) as ds:
        # Absent attribute (pre-M97 file shape) reads back as None.
        assert ds.genomic_runs["r0"].read_role is None


def test_read_role_round_trip_blocks(tmp_path: Path):
    from ttio.genomic.stream_writer import GenomicStreamWriter
    from ttio.spectral_dataset import SpectralDataset

    out = tmp_path / "blocks.tio"
    SpectralDataset.write_minimal(out, title="", isa_investigation_id="",
                                  runs={})
    run = _build_run(platform="ONT")
    with SpectralDataset.open(out, writable=True) as ds:
        w = GenomicStreamWriter(
            ds.study_group, "g",
            acquisition_mode=7, reference_uri=run.reference_uri,
            platform="ONT", sample_name="HG002",
            reference_chrom_seqs=run.reference_chrom_seqs,
            read_role="ont_ul",
        )
        w.append_batch(run)
        w.close()
    import h5py
    with h5py.File(out, "r") as f:
        rg = f["study/genomic_runs/g"]
        assert rg.attrs["read_role"] in ("ont_ul", b"ont_ul")
    with SpectralDataset.open(out) as ds:
        assert ds.genomic_runs["g"].read_role == "ont_ul"


# ---------------------------------------------------------------------------
# REF_DIFF_V2 slice_bytes — codec level
# ---------------------------------------------------------------------------

def test_slice_bytes_codec_byte_budget():
    """Non-uniform boundaries the fixed-count writer cannot produce,
    round-tripped by the unmodified decoder. Mirrors
    native/tests/test_ref_diff_v2_invariants.c test_m97_slice_bytes_policy."""
    n = 40
    ref = _ref_bytes(4096)
    lengths = [20 if r % 2 == 0 else 100 for r in range(n)]
    positions = np.array([r * 60 + 1 for r in range(n)], dtype=np.int64)
    offsets = np.zeros(n + 1, dtype=np.uint64)
    seqs = bytearray()
    for r in range(n):
        start = int(positions[r]) - 1
        read = bytearray(ref[start:start + lengths[r]])
        read[3] = ord("C") if read[3] == ord("A") else ord("A")
        seqs.extend(read)
        offsets[r + 1] = offsets[r] + lengths[r]
    total = int(offsets[n])
    cigars = [f"{L}M" for L in lengths]
    seq_arr = np.frombuffer(bytes(seqs), dtype=np.uint8)
    md5 = bytes(16)

    base = rdv2.encode(seq_arr, offsets, positions, cigars, ref, md5,
                       "m97", reads_per_slice=10_000)
    full = rdv2.encode(seq_arr, offsets, positions, cigars, ref, md5,
                       "m97", reads_per_slice=10_000, slice_bytes=total)
    assert full == base, "full budget must be byte-identical to default"

    budgeted = rdv2.encode(seq_arr, offsets, positions, cigars, ref, md5,
                           "m97", reads_per_slice=10_000, slice_bytes=200)
    assert _n_slices(budgeted) > 1
    assert budgeted != base

    # Non-uniform read counts: parse the 32-byte index entries
    # (num_reads is the u32 at entry offset 28).
    hdr = 38 + len("m97")
    counts = [
        struct.unpack_from("<I", budgeted, hdr + 32 * s + 28)[0]
        for s in range(_n_slices(budgeted))
    ]
    assert sum(counts) == n
    assert len(set(counts)) >= 2, f"uniform counts {counts}"

    out_seq, out_off = rdv2.decode(budgeted, positions, cigars, ref, n, total)
    assert out_seq.tobytes() == bytes(seqs)
    assert out_off.tobytes() == offsets.tobytes()


# ---------------------------------------------------------------------------
# REF_DIFF_V2 slice_bytes — through the writers
# ---------------------------------------------------------------------------

def test_slice_bytes_through_legacy_writer(tmp_path: Path):
    import h5py
    from ttio.spectral_dataset import SpectralDataset

    blob_path = "study/genomic_runs/r0/signal_channels/sequences/refdiff_v2"

    plain = _build_run(opt_legacy_whole_channel=True)
    out = _write_run(tmp_path, plain, "default.tio")
    with h5py.File(out, "r") as f:
        assert _n_slices(bytes(f[blob_path][()])) == 1

    budgeted = _build_run(opt_legacy_whole_channel=True,
                          ref_diff_slice_bytes=100)
    out2 = _write_run(tmp_path, budgeted, "budget.tio")
    with h5py.File(out2, "r") as f:
        # 100-base budget over 40 x 10 bp reads -> 4 slices of 10.
        assert _n_slices(bytes(f[blob_path][()])) == 4

    # The budgeted file round-trips through the ordinary reader.
    with SpectralDataset.open(out2) as ds:
        run = ds.genomic_runs["r0"]
        ref = _ref_bytes()
        for i in range(N):
            start = i * 20
            expect = ref[start:start + READ_LEN].decode("ascii")
            assert run[i].sequence == expect


def test_slice_bytes_through_blocks_writer(tmp_path: Path):
    import h5py
    from ttio.genomic.stream_writer import GenomicStreamWriter
    from ttio.spectral_dataset import SpectralDataset

    out = tmp_path / "blocks.tio"
    SpectralDataset.write_minimal(out, title="", isa_investigation_id="",
                                  runs={})
    run = _build_run()
    with SpectralDataset.open(out, writable=True) as ds:
        w = GenomicStreamWriter(
            ds.study_group, "g",
            acquisition_mode=7, reference_uri=run.reference_uri,
            platform="PacBio HiFi", sample_name="HG002",
            reference_chrom_seqs=run.reference_chrom_seqs,
            ref_diff_slice_bytes=100,
        )
        w.append_batch(run)
        w.close()
    with h5py.File(out, "r") as f:
        rg = f["study/genomic_runs/g"]
        idx = rg["blocks/index"][0]
        codec = int(idx["sequences_codec"])
        assert codec == int(Compression.REF_DIFF_V2)
        off, length = int(idx["sequences_off"]), int(idx["sequences_len"])
        blob = bytes(rg["signal_channels/sequences/data"][off:off + length])
        assert blob[:4] == b"RDF2"
        assert _n_slices(blob) == 4
